tsl/src/continuous_aggs/invalidation.c

```diff
@@ -423,49 +585,54 @@ clear_cagg_invalidations_for_refresh(const CaggInvalidationState *state,
 {
 	ScanIterator iterator;
 	int32 cagg_hyper_id = state->cagg.data.mat_hypertable_id;
+	Invalidation mergedentry;
 
-	/* The scanner defaults to SnapshotSelf, which would show the
-	 * modifications we do (new and updated tuples) within the scan loop
-	 * below. Therefore, we use our own snapshot here. */
+	invalidation_entry_reset(&mergedentry);
 	cagg_invalidations_scan_by_hypertable_init(&iterator, cagg_hyper_id, RowExclusiveLock);
-	iterator.ctx.snapshot = state->snapshot;
+
+	/* Must use an up-to-date snapshot to see inserts done after processing
+	 * the hypertable invalidation log */
+	iterator.ctx.snapshot = RegisterSnapshot(GetLatestSnapshot());
 	MemoryContextReset(state->per_tuple_mctx);
 
 	/* Process all invalidations for the continuous aggregate */
 	ts_scanner_foreach(&iterator)
 	{
-		Form_continuous_aggs_materialization_invalidation_log form;
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
 		MemoryContext oldmctx;
-		InvalidationResult result;
-		bool should_free;
-		HeapTuple tuple;
+		Invalidation logentry;
 
 		oldmctx = MemoryContextSwitchTo(state->per_tuple_mctx);
-		tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
-		form = (Form_continuous_aggs_materialization_invalidation_log) GETSTRUCT(tuple);
-		result = cut_invalidation_along_refresh_window(state,
-													   form->modification_time,
-													   form->lowest_modified_value,
-													   form->greatest_modified_value,
-													   refresh_window,
-													   &tuple->t_self);
-
-		switch (result)
+		invalidation_entry_set_from_cagg_invalidation(&logentry, ti);
+
+		if (!IS_VALID_INVALIDATION(&mergedentry))
+			mergedentry = logentry;
```

Comment:
This case is the first log entry and the entry should be removed or cut if inside the refresh window. My understanding is that it will happen on a future iteration, when try to merge fails or after the loop. Do I understand the logic correctly?